### PR TITLE
[s] [READY FOR REVIEW] ITS BACK : ATMOSPHERICS MULTILAYER PIPES!

### PR DESCRIPTION
**Pull request by silicons**
Description/images from testing/Changelog to come when I'm not deprived of sleep.
TL;DR:
Autismos

### CODE REVIEW

**Comment by Cyberboss** [comment]
Why tho?

**Comment by ma44** [comment]
Engineering points powergaming

**Comment by kevinz000** [comment]
@Cyberboss More compact autismospherics, and players get to screw around more when their creations aren't limited to using one tile per pipe.

**Comment by Supermichael777** [comment]
pipes are light tiles are expensive

**Comment by GunHog** [comment]
I am afraid of the insane contraptions that will spring up from this!

**Comment by kevinz000** [comment]
Ok let me put this in a different way
Why.. NOT? :^)

**Comment by Kraseo** [comment]
ehhehheeehheheeeee

**Comment by Dorsisdwarf** [comment]
It looks a little messy but I've wanted this for ages

**Comment by Incoming5643** [comment]
Air is already the most expensive thing on the station, I don't think we can really afford to make it more costly.

**Comment by RemieRichards** [comment]
No offence, but you are not the person I would trust to do this.

**Comment by Supermichael777** [comment]
How does it react to ventcrawling?

**Comment by RemieRichards** [comment]
@Supermichael777 Amusingly, both multipipes and ventcrawling were both written by the same /vg/ dev, ComicIronic. so they should work fine together.

**Comment by Dorsisdwarf** [comment]
I think they meant like, "how do you steer down the five-way junction" remie

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+/obj/machinery/atmospherics/pipe/layer_manifold/relaymove(mob/living/user, dir)
+	if(initialize_directions & dir)
+		return ..()
+	if((NORTH|EAST) & dir)
+		user.ventcrawl_layer = Clamp(user.ventcrawl_layer + 1, PIPING_LAYER_MIN, PIPING_LAYER_MAX)
+	if((SOUTH|WEST) & dir)
+		user.ventcrawl_layer = Clamp(user.ventcrawl_layer - 1, PIPING_LAYER_MIN, PIPING_LAYER_MAX)
+	user << "You align yourself with the [user.ventcrawl_layer]\th output."
+
+		/*
```
Comment (RemieRichards) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: dont include commented code.

**Comment by RemieRichards** [comment]
My response still works, the developer wrote both, and considered it.
What happens is you have the ventcrawling layer of the pipe you're in, and when you go through one of these multi-layer things, you take its layer, I'm not sure how this allows you to choose which one to go through though.

**Comment by DanteRU0312** [comment]
Finally. I'm so sick of the situations like when you want to lay pipes for some project, and you can't because there are scrubber pipes under the floor.

**Comment by FantasticFwoosh** [comment]
This'll be fun when i buckle someone to four different pipes across a single tile

**laughs in pipe**

**Comment by RemieRichards** [comment]
@FantasticFwoosh you can only be buckled to one thing, regardless of pipes.

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm (hunk 1 above)
Comment (kevinz000) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: Wip

**Comment by SgtxRootbeer** [comment]
I can now add an absurd amount of scrubbers to my art-closet supermatter engine

**Comment by FantasticFwoosh** [comment]
Well that is until somebody makes handcuffs localised to all limb slots to some extent to be chained by arms & legs to pipes wriggling out of each individually.

**Comment by Anonmare** [comment]
**Fear**

**Comment by CPTANT** [comment]
Do we really need 5 pipes on a tile though?

I think setting the limit to 3 would still provide greatly improved compactness for atmos projects, while also reducing the clutter immensely (compared to 5 pipes per tile).

**Comment by GunHog** [comment]
Let them have all five! Make it look like an ENGINE!

**Comment by kevinz000** [comment]
@CPTANT it only gets cluttered if players make it as I doubt mappers will use give tiles per tile much. However if this is merged after I finish it this would be an invitation to make distribution and waste lines share a tile :P

**Comment by CPTANT** [comment]
Just curious, but how do you determine what your pipe connects to when you wrench it in place?

**Comment by Cyberboss** [comment]
I'll be alright with this as long as we promise to never map these into stations

**Comment by PKPenguin321** [comment]
This is cool but you picked the worst possible example image

**Comment by PKPenguin321** [comment]
Like I know what this is so I like it but the uninitiated would look at that and think "gross I don't want that"

File: code/game/machinery/pipe/construction.dm [hunk 2]
```diff
@@ -261,16 +281,33 @@ var/global/list/pipeID2State = list(
...
 	qdel(src)
+
+/obj/item/pipe_meter/dropped()
+	..()
+	if(loc)
+		setAttachLayer(piping_layer)
+
+/obj/item/pipe_meter/proc/setAttachLayer(new_layer = PIPING_LAYER_DEFAULT)
+	piping_layer = new_layer
+	src.pixel_x = (new_layer - PIPING_LAYER_DEFAULT) * PIPING_LAYER_P_X
```
Comment (optimumtact) [inline code/game/machinery/pipe/construction.dm]: src

**Comment by Shadowlight213** [comment]
When can we do a config option to switch atmos mode between linda and zas? :^)

**Comment by RemieRichards** [comment]
Never, because LINDA > ZAS

**Comment by kevinz000** [comment]
@Shadowlight213 when you code it :3c

**Comment by Shadowlight213** [comment]
@RemieRichards is it even possible to do zas explosive decompression with linda though?

**Comment by kevinz000** [comment]
This is not dead I'll update in a bit.

**Comment by kevinz000** [comment]
so apparently pipes are connecting to other layers and other bullshit
time to see what's up...

**Comment by kevinz000** [comment]
ignore the laser stuff wrong branch

**Comment by kevinz000** [comment]
This is not dead I'm just to busy RN

**Comment by PKPenguin321** [comment]
Screee

**Comment by kevinz000** [comment]
SHOULD work properly this time, but don't fullmerge yet!

**Comment by MrStonedOne** [comment]
Suggestion: allow multiple pipes on a tile with this, but only allow one component

**Comment by CPTANT** [comment]
Perhaps introduce new components that allow multi pipe input/output?

**Comment by PJB3005** [comment]
> He didn't even ping Comic

@ComicIronic look at this heresy tbh.

**Comment by RemieRichards** [comment]
@PJB3005 of course he didn't? I doubt Kevinz even knows who Comic is.

**Comment by ComicIronic** [comment]
You navigate the manifold layer using the directions the manifold doesn't point in i.e. in a (west) 1-5 (east) manifold, north makes you move up a layer, and south makes you move down.

You should be able to change the number of layers by just editing PIPING_LAYER_MIN, PIPING_LAYER_DEFAULT, and PIPING_LAYER_MAX (and the manifold sprite). Everything else assumed only those defines, last I remember. 

Could do single-tile components using pipeflags, but I never set it up. You'd have to ask the current admins about what effect that has on the game.

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	findAllConnections()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/relaymove(mob/living/user, dir)
+	if(initialize_directions & dir)
+		return ..()
+	if((NORTH|EAST) & dir)
+		user.ventcrawl_layer = Clamp(user.ventcrawl_layer + 1, PIPING_LAYER_MIN, PIPING_LAYER_MAX)
+	if((SOUTH|WEST) & dir)
+		user.ventcrawl_layer = Clamp(user.ventcrawl_layer - 1, PIPING_LAYER_MIN, PIPING_LAYER_MAX)
+	user << "You align yourself with the [user.ventcrawl_layer]\th output."
```
Comment (ComicIronic) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: My commit was from before to_chat was a thing. Don't forget to change this.

Comment (PJB3005) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: TG doesn't have `to_chat`

Comment (RemieRichards) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: we do.

Comment (ComicIronic) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: From a quick ctrl+F, I believe they do.

Comment (PJB3005) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: Oh, since 9 days ago.

**Comment by Cyberboss** [comment]
Not to be a Debbie Downer but I feel like this is just gonna make pipes 10x
more broken than they already are.

**Comment by kevinz000** [comment]
@ComicIronic oh yeah pipes keep runtiming because somehow their parent pipenet datum gets deleted, and I'm in pain because I don't know atmoscode so I'm trying to decipher it . reee.

**Comment by MrStonedOne** [comment]
@kevinz000 me, @Aranclanos and @duncathan have worked hard to keep atmo pipes a symphony of carefully placed checks and creations and deletions of parents.

The runtimes are usually from pipes and or pipenets getting deleted that are still referenced by other pipe(nets) or something not being initialized correctly at the correct moment. (there are two points of initialization for pipes, when they are created, and when they are connected, and different things happen at different times.)

The best way to debug issues is giving pipes a log variable, outputting text to it for operations that mess with the parent value(s), (explaining whats going on and why) and then putting an if by the runtime location and outputting that log so you can see where the breakdown in logic is.

(Infact, I should make this a system that is built into the runtime tracking system)

**Comment by GunHog** [comment]
**Atmos slays yet another coder!**

**Comment by Jalleo** [comment]
Not really slay just produce a headache.

**Comment by kevinz000** [comment]
MWAHAHAHAHA ok the pipenets work fine now I just need to fix the fucking pipe overlays as they keep showing up as unconnected visually when they're conencted.

**Comment by SgtxRootbeer** [comment]
Get ready, get set... And freeze!

**Comment by GunHog** [comment]
> Get ready, get set... And freeze!

This is a pre-freeze PR.

**Comment by SgtxRootbeer** [comment]
Yeah, but the maintainers will prolly look at it and think "work on this was stopped pre-freeze, and resumed post-freeze". So I will be surprised to see how the fuzzy logic works out.

**Comment by kevinz000** [comment]
@SgtxRootbeer they let goof put guncrafting on the list last time.

**Comment by ExcessiveUseOfCobblestone** [comment]
>  players get to screw around more when their creations aren't limited to using one tile per pipe.

Obvious Non-Atmos Player detected!!!

**Comment by kevinz000** [comment]
alright finally this shit is done unless i missed some critical runtime...

**Comment by MrStonedOne** [comment]
`""""""""""""""""ready for review""""""""""""""""`

**Comment by kevinz000** [comment]
OK I FIXED MY SHIT BRING THE REVIEW HAMMER IM READY.

**Comment by PJB3005** [comment]
This is the best April's fools PR.

**Comment by duncathan** [comment]
I disagree the way this is currently set up
I will do a full review later

**Comment by optimumtact** [comment]
I just don't know about this, piping is already quite a complex system that runtimes consistently and we're gonna go on and add more complexity?

@duncathan I guess if you're happy with it then I'm happy to consider it but it is concerning.

**Comment by ComicIronic** [comment]
The actual complexity from the layering system isn't very high, unless your pipe connection code isn't particularly robust. The only special behaviour comes from the layer manifold, and it's nothing more special than a 4-fold with 2 more connections. Actual piping code is about the same.

**Comment by kevinz000** [comment]
@ComicIronic @duncathan @optimumtact when I did this the only hard part was the pipe layer manifold, as it took a long time for me to figure out how to get it to not fuck up pipenets. Now I just basically copypasted the normal code to nullify and connect nodes, only it does so with up to 10 connections (5 on each side) instead of just 1. The only other code that was runtiming was with components but I fixed that when I realized I accidently removed initialize directions checking, but yeah, the layer manifold is just a regular pipe but its code does whatever on 10 pipes instead of up to 4.

**Comment by optimumtact** [comment]
admitting you did copypaste code is not a way to make me positive about the pr

File: code/game/machinery/pipe/construction.dm [hunk 4]
```diff
@@ -261,16 +282,33 @@ GLOBAL_LIST_INIT(pipeID2State, list(
...
 		return ..()
-	if(!locate(/obj/machinery/atmospherics/pipe, src.loc))
-		to_chat(user, "<span class='warning'>You need to fasten it to a pipe!</span>")
+	var/obj/machinery/atmospherics/pipe/pipe
+	for(var/obj/machinery/atmospherics/pipe/P in src.loc)
+		if(P.piping_layer == piping_layer)
+			pipe = P
+			break
+	if(!pipe)
+		user << "<span class='warning'>You need to fasten it to a pipe!</span>"
```
Comment (optimumtact) [inline code/game/machinery/pipe/construction.dm]: to chat

File: code/game/machinery/pipe/construction.dm [hunk 5]
```diff
@@ -261,16 +282,33 @@ GLOBAL_LIST_INIT(pipeID2State, list(
...
 
 /obj/item/pipe_meter/attackby(obj/item/weapon/W, mob/user, params)
 	..()
 
 	if (!istype(W, /obj/item/weapon/wrench))
 		return ..()
-	if(!locate(/obj/machinery/atmospherics/pipe, src.loc))
-		to_chat(user, "<span class='warning'>You need to fasten it to a pipe!</span>")
+	var/obj/machinery/atmospherics/pipe/pipe
+	for(var/obj/machinery/atmospherics/pipe/P in src.loc)
```
Comment (optimumtact) [inline code/game/machinery/pipe/construction.dm]: \>src

File: code/game/machinery/pipe/construction.dm [hunk 6]
```diff
@@ -261,16 +282,33 @@ GLOBAL_LIST_INIT(pipeID2State, list(
...
+		if(P.piping_layer == piping_layer)
+			pipe = P
+			break
+	if(!pipe)
+		user << "<span class='warning'>You need to fasten it to a pipe!</span>"
 		return 1
-	new/obj/machinery/meter( src.loc )
+	new/obj/machinery/meter( src.loc , pipe)
```
Comment (optimumtact) [inline code/game/machinery/pipe/construction.dm]: src

File: code/game/objects/items/weapons/RPD.dm [hunk 7]
```diff
@@ -468,6 +478,13 @@ GLOBAL_LIST_INIT(RPD_recipes, list(
 		p_flipped = text2num(href_list["flipped"])
 		show_menu(usr)
 
+	if(href_list["setlayer"])
+		if(!isnum(href_list["setlayer"]))
+			piping_layer = text2num(href_list["setlayer"])
```
Comment (optimumtact) [inline code/game/objects/items/weapons/RPD.dm]: check is num again and bail if it's still not a valid number

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 8]
```diff
@@ -123,6 +153,11 @@ Pipelines + Other Objects -> Pipe network
 	return
 
 /obj/machinery/atmospherics/attackby(obj/item/weapon/W, mob/user, params)
+	if(istype(W, /obj/item/pipe)) //lets you autodrop
+		var/obj/item/pipe/pipe = W
+		user.drop_item(pipe)
+		pipe.setPipingLayer(piping_layer) //align it with us
+		return 1
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/atmosmachinery.dm]: true

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 9]
```diff
@@ -252,7 +285,7 @@ Pipelines + Other Objects -> Pipe network
 				var/list/pipenetdiff = returnPipenets() ^ target_move.returnPipenets()
 				if(pipenetdiff.len)
 					user.update_pipe_vision(target_move)
-				user.loc = target_move
+				user.forceMove(target_move)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/atmosmachinery.dm]: :heart:

File: code/modules/atmospherics/machinery/other/meter.dm [hunk 10]
```diff
@@ -29,8 +28,17 @@
 
 /obj/machinery/meter/Initialize(mapload)
 	..()
-	if (mapload && !target)
-		src.target = locate(/obj/machinery/atmospherics/pipe) in loc
+	if(mapload && !target)
+		for(var/obj/machinery/atmospherics/pipe/pipe in src.loc)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/other/meter.dm]: src

File: code/modules/atmospherics/machinery/other/meter.dm [hunk 11]
```diff
@@ -29,8 +28,17 @@
...
+	if(mapload && !target)
+		for(var/obj/machinery/atmospherics/pipe/pipe in src.loc)
+			if(pipe.piping_layer == target_layer)
+				target = pipe
+				setAttachLayer(pipe.piping_layer)
+				break
+
+/obj/machinery/meter/proc/setAttachLayer(var/new_layer)
+	target_layer = new_layer
+	src.pixel_x = (new_layer - PIPING_LAYER_DEFAULT) * PIPING_LAYER_P_X
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/other/meter.dm]: src

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+/obj/machinery/atmospherics/pipe/layer_manifold/New()
+	..()
+	volume = 260
+
+/obj/machinery/atmospherics/pipe/layer_manifold/proc/nullifyAllNodes()
+	var/list/obj/machinery/atmospherics/needs_nullifying = get_all_connected_nodes()
+	front_nodes = null
+	back_nodes = null
+	nodes = list()
+	for(var/obj/machinery/atmospherics/A in needs_nullifying)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: can this carry any other type? if not just do a typeless loop please

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+		A.disconnect(src)
+		if(A)
+			A.build_network()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/proc/get_all_connected_nodes()
+	var/list/obj/machinery/atmospherics/all_connected = list()
+	for(var/obj/machinery/atmospherics/I in front_nodes)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: if these are always going to be the same type please typeless loop

Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: or a subtype of that type

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+	var/list/obj/machinery/atmospherics/returnlist = list()
+	for(var/obj/machinery/atmospherics/A in all_connected)
+		returnlist += all_connected[A]
+	return returnlist
+
+/obj/machinery/atmospherics/pipe/layer_manifold/Destroy()
+	nullifyAllNodes()
+	..()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/update_icon()	//HEAVILY WIP FOR UPDATE ICONS!!
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: if it's WIP why have you marked this pr as ready?, remove this comment

File: code/modules/atmospherics/machinery/pipes/layermanifold.dm [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	piping_layer = PIPING_LAYER_DEFAULT
+
+/obj/machinery/atmospherics/pipe/layer_manifold/pipeline_expansion()
+	return get_all_connected_nodes()
+
+/obj/machinery/atmospherics/pipe/layer_manifold/disconnect(obj/machinery/atmospherics/reference)
+	if(istype(reference, /obj/machinery/atmospherics/pipe))
+		var/obj/machinery/atmospherics/pipe/P = reference
+		P.destroy_network()
+	while(reference in get_all_connected_nodes())
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/layermanifold.dm]: uhm why not get this as a list once and then loop over it? this is rebuilding a list over and over which is pretty ugly, is there no better way to do this?

why not book keep an allnodes list (updated whenever somethign is added/remove to nodes/frontnodes or backnodes) and then use that

File: code/modules/atmospherics/machinery/pipes/pipes.dm [hunk 16]
```diff
@@ -24,6 +24,16 @@
 	if(oldN)
 		oldN.build_network()
 
+/obj/machinery/atmospherics/pipe/destroy_network()
+	if(parent)
+		qdel(parent)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/pipes.dm]: QDEL_NULL

File: code/modules/atmospherics/machinery/pipes/pipes.dm [hunk 17]
```diff
@@ -56,14 +66,13 @@
...
-	if(!parent)
-		parent = new /datum/pipeline()
-		parent.build_pipeline(src)
-
 /obj/machinery/atmospherics/pipe/attackby(obj/item/weapon/W, mob/user, params)
 	if(istype(W, /obj/item/device/analyzer))
 		atmosanalyzer_scan(parent.air, user)
+	if(istype(W, /obj/item/pipe_meter))
+		var/obj/item/pipe_meter/meter = W
+		user.drop_item(meter, src.loc)
```
Comment (optimumtact) [inline code/modules/atmospherics/machinery/pipes/pipes.dm]: src src everywehre

**Review by optimumtact — CHANGES_REQUESTED** [review]
Style nits and some functionality questions

**Comment by duncathan** [comment]
I have spoken with Kevinz about this. This PR will not go through but multipipes will be redone from scratch sometime after the freeze after I finish rewriting pipenets.

**Pull request closed without merging**